src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs

```diff
@@ -2,20 +2,62 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
 namespace System.Globalization
 {
     internal static partial class GlobalizationMode
     {
         private static bool GetInvariantSwitchValue() =>
             GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
 
+        private static bool TryGetAppLocalIcuSwitchValue(out ReadOnlySpan<char> version, out ReadOnlySpan<char> icuSuffix)
+        {
+            icuSuffix = default;
+
+            if (!TryGetStringValue("System.Globalization.AppLocalIcu", "DOTNET_SYSTEM_GLOBALIZATION_APPLOCALICU", out string? value))
+            {
+                version = default;
+                return false;
+            }
+
+            // Custom built ICU can have a suffix on the name, i.e: libicuucmyapp.so.67.1
+            // So users would set the runtime switch as: myapp:67.1
+
+            int indexOfSeparator = value.IndexOf(':', StringComparison.Ordinal);
+            if (indexOfSeparator != -1)
+            {
+                ReadOnlySpan<char> valueAsSpan = value.AsSpan();
+                icuSuffix = valueAsSpan.Slice(0, indexOfSeparator);
+
+                if (icuSuffix.Length > 20)
+                {
+                    Environment.FailFast($"The resolved \"{icuSuffix.ToString()}\" suffix from System.Globalization.AppLocalIcu switch has to be < 20 chars long.");
```

Comment:
This value is provided via an AppContext switch; we're going to fail fast if it's invalid?

Reply:
Yeah, based on the design issue that's the behavior we want: 

> If the app opt-in to use local app version of ICU and failed to load it, we'll fail the app and fail fast.

That is what we also do when trying to load ICU from the system and can't find it:
https://github.com/dotnet/runtime/blob/59593640865d1c5869bfc72a27f623318411cb14/src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs#L24

Also in the native code we fail fast if we can't find exported symbols:
https://github.com/dotnet/runtime/blob/master/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L336